Ember upgrade v3.28 LTS to v4.12 LTS

benjay10 opened the pull request:
Upgrade to Ember v4.12 LTS from v3.28 LTS.

Meanwhile, I was able to fix some deprecations and rewrite some bits to hopefully future proof them.

benjay10 committed:
Remove package overrides

These package overrides where not really necessary.

benjay10 committed:
Resolve first deprecation about wormhole

benjay10 committed:
v3.28.5...v4.12.2

benjay10 committed:
Upgrades and removals of dependencies

benjay10 committed:
Deprecation fix: explicit JSONAPIAdapter

benjay10 committed:
Deprecation fix: inverse and async options on models

benjay10 committed:
Repair missing ember-power-select

benjay10 committed:
Deprecation fix: replace firstObject

This meant adding an extra function for the login to resolve promises
and proxies.

benjay10 committed:
Deprecation fix: find → findRecord

benjay10 committed:
Removed old fix that is no longer needed

benjay10 committed:
Remove .gitkeep files that are not needed

benjay10 committed:
Rewrite tasks with moders syntax

This allows us to remove the ember-concurrency-decorators dependency.

benjay10 committed:
Replace <AuLink> with <a> to stop warning

There was a warning about a missing model or missing route because these
where not supplied to these links. That is because these links are not
used in the traditional sense of a navigation link but more as a button.

benjay10 committed:
Bump Docker builder version

benjay10 committed:
Remove GitHub workflow

benjay10 committed:
Remove Drone CI

Windvis commented on .github/workflows/ci.yml:
This file isn't needed.

Windvis commented on app/routes/mock-login.js:
```diff
@@ -19,7 +19,7 @@ export default class MockLoginRoute extends Route {
     if (params.gemeente) filter.gebruiker = { achternaam: params.gemeente };
     return this.store.query('account', {
-      include: 'gebruiker.bestuurseenheden',
+      include: 'gebruiker,gebruiker.bestuurseenheden',
```
I don't think this change was needed? Or do you think it's clearer this way?

benjay10 replied:
I think it looks clearer, but I thought you had to include it to get the attributes. Including `gebruiker.bestuurseenheden` will load the attributes for the bestuurseenheid but not necessarily for gebruiker, or is that implied?

Windvis replied:
Yea, it implicitly does that (AFAIK), but I can understand that adding it explicitly is clearer.

Windvis commented on app/styles/app.css:
This file shouldn't be needed, but you might need to tweak the lint script to allow it to not find any css files. I think I've done it in some other project before I switched the setup to support sass linting.

https://github.com/lblod/frontend-lpdc/commit/c9ad866d67ad66d4e6ec4a8307cd29b6ed19efec#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L19

benjay10 replied:
This file came as part of the automatic Ember upgrade, just like the GitHub Workflow stuf and a few other things. Will remove it.

Windvis commented on app/templates/forms/edit.hbs:
```diff
@@ -16,24 +16,24 @@
-                @iconAlignment="left"
-                @skin="primary"
-                class="au-c-list-navigation__link {{if this.formVisible "active"}}"
+              <a
```
Would a `AuButton` work?

benjay10 replied:
No, I tried. The styling is wrong. Being a button, it is smaller and doesn't get the margins and padding it would get as a link.

Windvis replied:
Well, then I would fix the styles 😄.

benjay10 replied:
What about this (see commit https://github.com/lblod/frontend-meldingsplichtige-api/pull/32/commits/6f8a3d7230a4ba0a6d3823d5e4593cb184edbd41)?

Added a new class with overrides to style the button just like the other navigation links (apart from the outline, but that is fine).

Windvis replied:
I like it 👍

Windvis commented on app/controllers/mock-login.js:
```diff
@@ -11,24 +12,32 @@ export default class MockLoginController extends Controller {
+  @action
+  async login(loginF, account) {
+    const accountId = account.id;
+    const gebruiker = account.gebruiker;
```
The relationship is async so it needs to be awaited first.
```suggestion
    const gebruiker = await account.gebruiker;
```

Another option is to make it `sync` instead, the (mock) users usually are fetched with includes I think? That would make things a lot simpler.

benjay10 replied:
Ah makes sense. I'll make it synchronous. Then we also don't need await and the gebruiker is no longer a proxy which means we can solve the next problem as well.

Windvis commented on app/controllers/mock-login.js:
```diff
@@ -11,24 +12,32 @@ export default class MockLoginController extends Controller {
+  async login(loginF, account) {
+    const accountId = account.id;
+    const gebruiker = account.gebruiker;
+    const bestuurseenheden = await gebruiker.get('bestuurseenheden');
```
if `account.gebruiker` is awaited, `.get` shouldn't be needed here.

benjay10 replied:
It was necessary. The console error told me this was an exceptional case where this was needed. Although after awaiting the gebruiker, it is no longer needed.

benjay10 committed:
Use Woodpecker CI

benjay10 committed:
Make relation between Account and Gebruiker sync

These models are always loaded at the same time, so it would make sense
to do so. It is no longer a proxy and doesn't need `.get()`.

benjay10 committed:
Remove unneeded CSS file

benjay10 committed:
Rollback on sync relation

There is a warning after logging in that warns about a known-to-be-empty
relation. This is because when the current session is loaded, the
gebruiker data is not included. Login is done via ember-simple-auth
and adding gebruiker in the includes will be more trouble than marking
this relation sync again and using await.

benjay10 committed:
Linting: Remove autofocus and add empty role

benjay10 committed:
Changed <a> into <AuButton>

This is more semantically correct. Using a button for actions like
showing and hiding content. This however meant that some styling
overrides are needed.

benjay10 committed:
Ignore: silly styling in README to trigger a CI build

benjay10 committed:
Synchronise package-lock.json

benjay10 committed:
Removed unmaintained and unused dependency

This would hopefully fix that `npm ci` problem. This dependency was not
used and is unmaintained since 2016.

benjay10 committed:
Allow linting to fail, but not the whole CI pipeline

Windvis commented on .woodpecker/.test-build.yml:
```diff
@@ -0,0 +1,19 @@
+      - npm run lint:js
+      - npm run lint:hbs
+      - npm run lint:dependency
+    failure: ignore
```
Please don't 😄. Looking at the CI faillures it doesn't seem impossible to fix?

I would add eslint ignore comments to the mixins since there is no way around those at the moment.

For the dependency lint issues:
- ember-modifiers: use overrides to force v4 (assuming there are no deprecation warnings shown in app)
- ember-concurrency: either use overrides to v3 or use v2 in the app instead. Appuniversum supports both 2 and 3 so it should follow the app version (and if not, a npm dedupe will probably fix it).

Pull requests to add support for these versions in the addons is the best fix but the overrides can be used as a workaround for now.

benjay10 replied:
Pretty sure I have seen linting (especially dependency linting) to be allowed to fail in other repositories with Drone. That's why I did that here too.

benjay10 committed:
Revert "Allow linting to fail, but not the whole CI pipeline"

This reverts commit 29e1c01c1bb98098c676cf29122a2dfca3b73bf7.

benjay10 committed:
Added overrides to resolve dependency linting issues
